components/service/glean/src/test/java/mozilla/components/service/glean/GleanTest.kt

```diff
@@ -513,4 +501,112 @@ class GleanTest {
+            lifetime = Lifetime.Ping,
+            name = "string_metric",
+            sendInPings = listOf("default")
+        )
+
+        stringMetric.set("TEST VALUE")
+        assertTrue(stringMetric.testHasValue())
+
+        Glean.setUploadEnabled(false)
+        assertFalse(stringMetric.testHasValue())
+        stringMetric.set("TEST VALUE")
+        assertFalse(stringMetric.testHasValue())
+
+        Glean.setUploadEnabled(true)
+        assertFalse(stringMetric.testHasValue())
+        stringMetric.set("TEST VALUE")
+        assertTrue(stringMetric.testHasValue())
+    }
+
+    @Test
+    fun `Core metrics should be cleared and restored when disabling and enabling uploading`() {
+        assertTrue(GleanInternalMetrics.os.testHasValue())
+
+        Glean.pingMaker.getPingSeq("custom")
+
+        Glean.setUploadEnabled(false)
+        assertFalse(GleanInternalMetrics.os.testHasValue())
+        assertEquals(0, Glean.pingMaker.getPingSeq("custom"))
+
+        Glean.setUploadEnabled(true)
+        assertTrue(GleanInternalMetrics.os.testHasValue())
+    }
+
+    @Test
+    fun `clientId is managed correctly when disabling and enabling metrics`() {
+        val originalClientId = GleanInternalMetrics.clientId.testGetValue()
+        assertNotEquals(GleanInternalAPI.KNOWN_CLIENT_ID, GleanInternalMetrics.clientId.testGetValue())
+
+        Glean.setUploadEnabled(false)
+        assertEquals(GleanInternalAPI.KNOWN_CLIENT_ID, GleanInternalMetrics.clientId.testGetValue())
+
+        Glean.setUploadEnabled(true)
+        assertNotEquals(GleanInternalAPI.KNOWN_CLIENT_ID, GleanInternalMetrics.clientId.testGetValue())
+        assertNotEquals(originalClientId, GleanInternalMetrics.clientId.testGetValue())
+    }
+
+    @Test
+    fun `firstRunDate is managed correctly when disabling and enabling metrics`() {
+        val originalFirstRunDate = GleanInternalMetrics.firstRunDate.testGetValue()
+
+        Glean.setUploadEnabled(false)
+        assertEquals(originalFirstRunDate, GleanInternalMetrics.firstRunDate.testGetValue())
+
+        Glean.setUploadEnabled(true)
+        assertEquals(originalFirstRunDate, GleanInternalMetrics.firstRunDate.testGetValue())
+    }
+
+    @Test
+    fun `disabling upload clears pending pings`() {
+        val stringMetric = StringMetricType(
+            disabled = false,
+            category = "telemetry",
+            lifetime = Lifetime.Ping,
+            name = "string_metric",
+            sendInPings = listOf("default")
+        )
+
+        stringMetric.set("TEST VALUE")
+
+        Pings.baseline.send()
+        assertEquals(1, Glean.pingStorageEngine.testGetNumPendingPings())
+
+        Glean.setUploadEnabled(false)
+        assertEquals(0, Glean.pingStorageEngine.testGetNumPendingPings())
+
+        Glean.setUploadEnabled(true)
+        assertEquals(0, Glean.pingStorageEngine.testGetNumPendingPings())
+    }
+
+    @Test
+    fun `clientId is set to known id when setting disabled from a cold start`() {
```

Comment:
Can we move the `clientId is managed correctly when disabling and enabling metrics` test closer to these 2? They are testing related things, so having them close by makes sense IMO.